core/finality-grandpa/src/finality_proof.rs

```diff
@@ -0,0 +1,410 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! GRANDPA block finality proof generation and check.
+//!
+//! Finality of block B is proved by providing:
+//! 1) valid headers sub-chain from the block B to the block F;
+//! 2) valid (with respect to proved authorities) GRANDPA justification of the block F;
+//! 3) proof-of-execution of the `grandpa_authorities` call at the block F.
+//!
+//! Since earliest possible justification is returned, the GRANDPA authorities set
+//! at the block F is guaranteed to be the same as in the block B (this is because block
+//! that enacts new GRANDPA authorities set always comes with justification). It also
+//! means that the `set_id` is the same at blocks B and F.
+//!
+//! The caller should track the `set_id`. The most straightforward way is to fetch finality
+//! proofs ONLY for blocks on the tip of the chain and track the latest known `set_id`.
+
+use std::collections::HashMap;
+
+use client::{
+	blockchain::Backend as BlockchainBackend,
+	error::{Error as ClientError, ErrorKind as ClientErrorKind, Result as ClientResult},
+	light::fetcher::RemoteCallRequest,
+};
+use codec::{Encode, Decode};
+use grandpa::BlockNumberOps;
+use runtime_primitives::generic::BlockId;
+use runtime_primitives::traits::{
+	NumberFor, Block as BlockT, Header as HeaderT, One,
+};
+use substrate_primitives::{AuthorityId, H256};
+
+use GrandpaJustification;
+
+/// Prepare proof-of-finality for the given block.
+///
+/// The proof is the serialized `FinalityProof` constructed using earliest known
+/// justification of the block. None is returned if there's no known justification atm.
+pub fn prove_finality<Block: BlockT, B, G>(
+	blockchain: &B,
+	generate_execution_proof: G,
+	block: Block::Hash,
+) -> ::client::error::Result<Option<Vec<u8>>>
+	where
+		B: BlockchainBackend<Block>,
+		G: Fn(&BlockId<Block>, &str, &[u8]) -> ClientResult<Vec<Vec<u8>>>,
+{
+	let block_id = BlockId::Hash(block);
+	let mut block_number = blockchain.expect_block_number_from_id(&block_id)?;
+
+	// early-return if we sure that the block isn't finalized yet
+	let info = blockchain.info()?;
+	if info.finalized_number < block_number {
+		return Ok(None);
+	}
+
+	// early-return if we sure that the block is NOT a part of canonical chain
+	let canonical_block = blockchain.expect_block_hash_from_id(&BlockId::Number(block_number))?;
+	if block != canonical_block {
+		return Err(ClientErrorKind::Backend(
+			"Cannot generate finality proof for non-canonical block".into()
+		).into());
+	}
+
+	// now that we know that the block is finalized, we can generate finalization proof
+
+	// we need to prove grandpa authorities set that has generated justification
+	let authorities_proof = generate_execution_proof(&block_id, "GrandpaApi_grandpa_authorities", &[])?;
```

Comment:
We should rather get a proof of this execution on the parent block -- it's undocumented at this point but what `grandpa_authorities` really tells you is which set of authorities is responsible for finalizing children of this block.

This is because when we do a set change at block X, `grandpa_authorities(at: X)` will give you the new set, but it's actually the old set who is responsible for finalizing X and completing the handoff.

Adding extra documentation to the `GrandpaApi` trait would be really helpful as well